src/Microsoft.AspNetCore.Server.IIS/Core/IISNativeApplication.cs

```diff
@@ -31,7 +32,17 @@ public void StopCallsIntoManaged()
             IntPtr requestContext,
             IntPtr shutdownContext)
         {
-            NativeMethods.HttpRegisterCallbacks(_nativeApplication, requestHandler, shutdownHandler, onAsyncCompletion, requestContext, shutdownContext);
+            NativeMethods.HttpRegisterCallbacks(_nativeApplication, requestHandler, shutdownHandler, onAsyncCompletion, requestContext, shutdownContext, out var fLoggingEnabled);
+
+            if (!fLoggingEnabled)
+            {
+                // By using the PipeOutputRedirection, after calling RegisterCallbacks,
+                // stdout and stderr will be redirected to NULL. However, if something wrote
+                // to stdout before redirecting, (like a Console.WriteLine during startup),
+                // we need to call Console.Set* to pick up the modified consoles outputs.
+                Console.SetOut(new StreamWriter(Console.OpenStandardOutput()));
```

Comment:
Use same arguments as https://github.com/dotnet/corefx/blob/master/src/System.Console/src/System/Console.cs#L120 as much as possible.

Reply:
Can OpenStandardOutput be anything but `NUL` at this point?

Reply:
For IISExpress, it wouldn't be `NUL`.

Reply:
We won't redirect anywhere in the case where we have actual terminal right?

Reply:
Yes. but I think some assumptions we have about tests logging to stdout will fail. I'll try and see what happens.